akka-actor-typed-tests/src/test/scala/akka/actor/typed/testkit/EventFilter.scala

```diff
@@ -0,0 +1,474 @@
+ * error messages.
+ *
+ * See the companion object for convenient factory methods.
+ *
+ * If the `occurrences` is set to Int.MaxValue, no tracking is done.
+ */
+abstract class EventFilter(occurrences: Int) {
+
+  /*
+   * these default values are just there for easier subclassing
+   */
+  protected val source: Option[String] = None
+  protected val message: Either[String, Regex] = Left("")
+  protected val complete: Boolean = false
+  @volatile // JMM does not guarantee visibility for non-final fields
+  private var todo = occurrences
+
+  import scala.collection.JavaConverters._
+
+  /**
+   * Apply this filter while executing the given code block. Care is taken to
+   * remove the filter when the block is finished or aborted.
+   */
+  def intercept[T](code: => T, eventsQueue: LinkedBlockingQueue[SubstituteLoggingEvent])(
+      implicit system: ActorSystem): T = {
+    //TODO add appender and clean it @see original TestEventListener
+    def leftToDo: Int = todo - eventsQueue.asScala.filter(each => matches(each)).size
+    val leeway = TestKitExtension(system).TestEventFilterLeeway
+    val result = code
+    if (!awaitDone(leeway, leftToDo))
+      if (leftToDo > 0)
+        throw new AssertionError(s"timeout ($leeway) waiting for $leftToDo messages on $this")
+      else
+        throw new AssertionError(s"received ${-leftToDo} excess messages on $this")
+
+    result
+  }
+
+  def awaitDone(max: Duration, leftToDo: => Int): Boolean = {
+    if (leftToDo != Int.MaxValue && leftToDo > 0) TestKit.awaitCond(leftToDo <= 0, max, noThrow = true)
+    leftToDo == Int.MaxValue || leftToDo == 0
+  }
+
+  /**
+   * This method decides whether to filter the event (<code>true</code>) or not
+   * (<code>false</code>).
+   */
+  protected def matches(event: LoggingEvent): Boolean
+
+  /**
+   * internal implementation helper, no guaranteed API
+   */
+  protected def doMatch(msg: Any) = {
+    val msgstr = if (msg != null) msg.toString else "null"
+    (message match {
+      case Left(s)  => if (complete) msgstr == s else msgstr.startsWith(s)
+      case Right(p) => p.findFirstIn(msgstr).isDefined
+    })
+  }
+}
+
+/**
+ * Facilities for selectively filtering out expected events from logging so
+ * that you can keep your test run’s console output clean and do not miss real
+ * error messages.
+ *
+ * '''Also have a look at the `akka.testkit` package object’s `filterEvents` and
+ * `filterException` methods.'''
+ *
+ * The source filters do accept `Class[_]` arguments, matching any
+ * object which is an instance of the given class, e.g.
+ *
+ * {{{
+ * EventFilter.info(source = classOf[MyActor]) // will match Info events from any MyActor instance
+ * }}}
+ *
+ * The message object will be converted to a string before matching (`"null"` if it is `null`).
+ */
+object EventFilter {
+
+  /**
+   * Create a filter for Error events. Give up to one of <code>start</code> and <code>pattern</code>:
```

Comment:
It seems to me — at least part of what it is supposed to do. I don't think that the description is complete though, as it can filter all level of messages. 

I added a mention to org.slf4j.event.LoggingEvent mention.